File: parser.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "lexer.h"
#include "utilities.h"
#include "parser.h"

/*
 * Directs the entire parse. Using recursive descent and shunting yard
 * algorithms. Currently parsing mathematical expressions.
 */
int parse(char **pos)
{
    int symbol_table_size = 512;
    symbol_table_t *symbol_table = init_symbol_table(symbol_table_size);
    while(parse_assignment(pos, symbol_table));

    //stuff
    stack *operators=malloc(sizeof(stack)), *operands=malloc(sizeof(stack));
    push(operators, init_element("sentinel"));
    if(parse_expression(pos, symbol_table, operators, operands))
        printf("Result: %s\n", pop(operands)->data);
    else
        printf("Parse failed :(\n");
    
    return 1;
}

/*
 * Initializes and places values in symbol table
 * assignment -> id = val;
 */
int parse_assignment(char **pos, symbol_table_t *symbol_table)
{
    char **lookahead = malloc(sizeof(char *));
    char *id;
    type_union_t val;
    token *tok;
    symbol_t *sym;
    stack *operators=malloc(sizeof(stack)), *operands=malloc(sizeof(stack));

    *lookahead = *pos;
    tok = malloc(sizeof(token));

    tok = scan(lookahead);
    if(tok && !strcmp(tok->token, "ID"))
    {
        id = strdup(tok->attribute);
        tok = scan(lookahead);
        if(tok && !strcmp(tok->token, "="))
        {
            push(operators, init_element("sentinel"));
            if(parse_expression(lookahead, symbol_table, operators, operands))
            {
                val.i = atoi(pop(operands)->data);
                if((sym = get_symbol_table(symbol_table, id)))
                    sym->val = val;
                else
                    insert_symbol_table(symbol_table, id, val); //Add the type here when you implement type checking

                tok = scan(lookahead);
                if(tok && !strcmp(tok->token, ";"))
                {
                    *pos = *lookahead;
                    return 1;
                }
            }
        }
    }

    return 0;
    
}


/*
 * expr -> expr {binary_op expr}
 */
int parse_expression(char **pos, symbol_table_t *symbol_table, stack *operators, stack *operands)
{
    char **lookahead = malloc(sizeof(char *));
    char **copy = malloc(sizeof(char *));
    token *tok;

    *lookahead = *pos;
    tok = malloc(sizeof(token));

    if(!parse_factor(lookahead, symbol_table, operators, operands))
        return 0;

    while(1)
    {
        *copy = *lookahead;
        if(!parse_binary_op(lookahead, symbol_table, operators, operands))
        {
            *lookahead = *copy;
            break;
        }

        if(!parse_factor(lookahead, symbol_table, operators, operands))
        {
            printf("Error parsing: missing variable or integer literal.\n");
            exit(0);
        }
    }
    
    if(!parse_expression_end(lookahead, operators, operands))
    {
        *lookahead = *copy;
        printf("Error parsing: expression could not be evaluated.\n");
        exit(0);
    }

    *pos = *lookahead;
    return 1;
}


/*
 * factor -> int | (expr) | unary_op factor
 */
int parse_factor(char **pos, symbol_table_t *symbol_table, stack *operators, stack *operands)
{
    char **lookahead = malloc(sizeof(char *));
    token *tok;
    symbol_t *sym;

    *lookahead = *pos;
    tok = malloc(sizeof(token));

    tok = scan(lookahead);
    if(tok)
    {
        if(!strcmp(tok->token, "INT"))
        {
             push(operands, init_element(tok->attribute));
             goto pass;
        }

        if(!strcmp(tok->token, "("))
        {
            push(operators, init_element("sentinel"));
            if(!parse_expression(lookahead, symbol_table, operators, operands))
                return 0;

            tok = scan(lookahead);
            if(tok && !strcmp(tok->token, ")"))
            {
                pop(operators);
                goto pass;
            }
        }

        if(!strcmp(tok->token, "-") || !strcmp(tok->token, "+"))
        {
            if(parse_unary_op(lookahead, symbol_table, operators, operands, tok->token))
                goto pass;
        }

        if(!strcmp(tok->token, "ID"))
        {
            sym = get_symbol_table(symbol_table, tok->attribute);
            if(sym)
            {
                push(operands, init_element(itoa(sym->val.i)));
                goto pass;
            }
            else
            {
                printf("Error parsing: uninitialized variable\n");
                exit(0);
            }
        }
    }

    return 0;

//Using a goto statement like a bad boy
pass:
    *pos = *lookahead;
    return 1;
}

/*
 * binary_op -> + | - | * | /
 */
int parse_binary_op(char **pos, symbol_table_t *symbol_table, stack *operators, stack *operands)
{
    char **lookahead = malloc(sizeof(char *));
    *lookahead = *pos;
    token *tok;
    element *top;
    int v1, v2;

    tok = scan(lookahead);
    if(tok &&
            (!strcmp(tok->token, "+") ||
             !strcmp(tok->token, "-")))
    {    
        top = peek(operators);
 
        //push current operator while waiting for next operand to evaluate
        if(!strcmp(top->data, "sentinel"))
        {
            push(operators, init_element(tok->token));
            *pos = *lookahead;
            return 1;
        }
        
        //while theres higher precednece operators on the stack perform them
        while(top &&
                (!strcmp(top->data, "*") || !strcmp(top->data, "/")))
        {
            pop(operators);
            v2 = atoi(pop(operands)->data);
            v1 = atoi(pop(operands)->data);

            !strcmp(top->data, "*") ?
                push(operands, init_element(itoa(v1*v2))) :
                push(operands, init_element(itoa(v1/v2)));

            top = peek(operators);
        }

        if(top &&
                (!strcmp(top->data, "+") || !strcmp(top->data, "-")))
        {
            pop(operators);
            v2 = atoi(pop(operands)->data);
            v1 = atoi(pop(operands)->data);

            !strcmp(top->data, "+") ?
                push(operands, init_element(itoa(v1+v2))) :
                push(operands, init_element(itoa(v1-v2)));

        }

        push(operators, init_element(tok->token));
        *pos = *lookahead;
        return 1;
    }


    if(tok &&
            !strcmp(tok->token, "*") ||
            !strcmp(tok->token, "/"))
    {
        top = peek(operators);

        if(!strcmp(top->data, "sentinel"))
        {
            push(operators, init_element(tok->token));
            *pos = *lookahead;
            return 1;
        }
        
        //perform greater or equal ops first
        while(top &&
                (!strcmp(top->data, "*") || !strcmp(top->data, "/")))
        {
            pop(operators);
            v2 = atoi(pop(operands)->data);
            v1 = atoi(pop(operands)->data);

            !strcmp(top->data, "*") ?
                push(operands, init_element(itoa(v1*v2))) :
                push(operands, init_element(itoa(v1/v2)));

            top = peek(operators);
        }

        push(operators, init_element(tok->token));
        *pos = *lookahead;
        return 1;
    }

    return 0;

}

/*
 * unary_op -> + | -
 */
int parse_unary_op(char **pos, symbol_table_t *symbol_table, stack *operators, stack *operands, char *prefix)
{
    char **lookahead = malloc(sizeof(char *));
    *lookahead = *pos;
    token *tok;
    element *top;

    tok = scan(lookahead);
    if(tok && !strcmp(tok->token, "INT"))
    {
        top = peek(operators);
        if(top &&
                (!strcmp(top->data, "*") ||
                 !strcmp(top->data, "/") ||
                 !strcmp(top->data, "sentinel")))
        {
            if(!strcmp(prefix, "-"))
                push(operands, init_element(itoa(-atoi(tok->attribute))));
            else
                push(operands, init_element(tok->attribute));
            
            *pos = *lookahead;
            return 1;
        }
    }

    return 0;
}

/*
 * When we reach the end of a mathematical expression we probably still have
 * operators and operands waiting on their respective stacks. This function
 * clears and evaluates them.
 */
int parse_expression_end(char **pos, stack *operators, stack *operands)
{
    int v1, v2;
    token *tok;
    char **lookahead = malloc(sizeof(char *));

    *lookahead = *pos;

    while(strcmp(peek(operators)->data, "sentinel"))
    {
        v2 = atoi(pop(operands)->data);
        v1 = atoi(pop(operands)->data);

        char op = *pop(operators)->data;
        switch(op)
        {
            case '+':
                push(operands, init_element(itoa(v1+v2)));
                *pos = *lookahead;
                break;

            case '-':
                push(operands, init_element(itoa(v1-v2)));
                *pos = *lookahead;
                break;
            
            case '*':
                push(operands, init_element(itoa(v1*v2)));
                *pos = *lookahead;
                break;
            
            case '/':
                push(operands, init_element(itoa(v1/v2)));
                *pos = *lookahead;
                break;
            
            default:
                printf("Error parsing: expected binary operator +, -, *, or /.\n");
                exit(0);
        }
    }

    return 1;
}

File: Parser.py
import sys
import Lexer as Lex

class Stack:
    def __init__(self):
        self.elements = []

    def push(self, element):
        self.elements.append(element)

    def pop(self):
        element = self.elements[len(self.elements)-1]
        del self.elements[len(self.elements)-1]
        return element

    def peek(self):
        if self.elements != []:
            return self.elements[len(self.elements)-1]

    def is_empty(self):
        return self.elements == []

    def __str__(self):
        return str([str(i) for i in self.elements])


class Parser:                                                     #can manually set the lexer's self.pos variable like passing in &pos in C
    def __init__(self, path):
        self.path = path
        self.file = open(path, 'r')
        self.code = self.file.read() + '$$'
        self.file.close()
        self.lexer = Lex.Lexer(self.code)
        self.operands = Stack()
        self.operators = Stack()
        self.operators.push(Lex.Token('sentinel', 'sentinel'))

    def parse(self):
        if self.parse_expr():
            tok = self.lexer.scan()
            if tok.token == "semicolon":
                print("Successfully parsed: {}".format(self.lexer.code[:self.lexer.pos]))
                self.shunting_yard(0, self.lexer.pos)
                return
        print("Failure")
        print(self.operands, self.operators)

    #expr -> factor {binary_op factor}
    def parse_expr(self):
        if not self.parse_factor():
            return 0

        while True:
            if not self.parse_binary_op():
                return 1

            if not self.parse_factor():
                return 0
        
    
    #factor -> int | (expr) | id | unary_op factor
    def parse_factor(self):
        pos_0 = self.lexer.pos
        tok = self.lexer.scan()

        if tok.token == "INT":
            self.operands.push(int(tok.att))
            return 1

        if tok.token == "ID":
            return 1

        if tok.att == "(":             #XXX supposed to push the sentinel value onto the stack of operators here
            if self.parse_expr():
                tok = self.lexer.scan()
                if tok.att == ")":
                    return 1

        #reset position in text file on failure
        self.lexer.pos = pos_0
        return 0

    #binary_op -> mult_op | add_op
    def parse_binary_op(self):
        pos_0 = self.lexer.pos
        tok = self.lexer.scan()

        if tok.token == "add_op":
            top = self.operators.peek()

            #wait for next operand to know if you should do something yet
            if top.token == 'sentinel':
                self.operators.push(tok)
                return 1

            while top.token == 'mult_op':
                self.operators.pop()
                v2 = self.operands.pop()
                v1 = self.operands.pop()
                if top.att == '*':
                    self.operands.push(v1 * v2)
                elif top.att == '/':
                    self.operands.push(v1 / v2)
                else:
                    print("Error in parse_binary_op")
                    sys.exit(0)
                top = self.operators.peek()

            if top.token == 'add_op':
                self.operators.pop()
                v2 = self.operands.pop()
                v1 = self.operands.pop()
                if top.att == '+':
                    self.operands.push(v1 + v2)
                elif top.att == '-':
                    self.operands.push(v1 - v2)
                else:
                    print("Error in parse_binary_op")
                    sys.exit(0)
                top = self.operators.peek()
            self.operators.push(tok)
            return 1

        if tok.att == 'mult_op':
            top = self.operators.peek()
            if top.att == 'sentinel':
                self.operators.push(tok)
                return 1

            while top.token == 'mult_op':
                self.operators.pop()
                v2 = self.operands.pop()
                v1 = self.operands.pop()
                if top.att == '*':
                    self.operands.push(v1 * v2)
                elif top.att == '/':
                    self.operands.push(v1 / v2)
                else:
                    print("Error in parse_binary_op")
                    sys.exit(0)
                top = self.operators.peek()

            self.operators.push(tok)
            return 1

        #reset position in text file on failure
        self.lexer.pos = pos_0
        return 0

    #unary_op -> + | -          Eventually *, &, etc
    def parse_unary_op(self):
        pass
"""
    def shunting_yard(self, start, end):
        sy_lexer = Lex.Lexer(self.code[start:end])

        tok = sy_lexer.scan()
        while tok.att != ';':
            print("current token: {}".format(tok))
            
            #push all numbers to operands stack
            if tok.token == 'INT':
                self.operands.push(int(tok.att))

            #open parens jsut get pushed to mark for close parens
            elif tok.att == '(':
                self.operators.push(Lex.Token('sentinel', 'sentinel'))

            '''
            #evaluate the inside of parentheses
            elif tok.att == ')':
                #get rid of the open paren
                self.operators.pop()

                #get the last operator seen from the stack
                top = self.operators.peek()

                #while we haven't seen  XXX: used to have "top.att != '(' and" at the front of if condition
                while tok.get_precedence() <= top.get_precedence():
                    top = self.operators.pop()
                    if top.att == 'sentinel':
                        break
                    print("top is: {}".format(top))
                    v2 = self.operands.pop()
                    v1 = self.operators.pop()
                    if top.att == '*':
                        self.operands.push(v1 * v2)
                    elif top.att == '/':
                        self.operands.push(v1 / v2)
                    elif top.att == '+':
                        self.operands.push(v1 + v2)
                    elif top.att == '-':
                        self.operands.push(v1 - v2)
                    else:
                        print("Error in shunting yard paren while loop")
                        sys.exit(0)
                self.operators.pop()'''
            elif tok.att == '*' or tok.att == '/' or tok.att == '+' or tok.att == '-':
                #XXX
                if self.operators.is_empty():
                    self.operators.push(tok)
                    tok = sy_lexer.scan()
                    continue

                top = self.operators.peek()
                #usual condition is that while operators isnt empty
                while not top.att == 'sentinel' and tok.get_precedence() <= top.get_precedence():
                    print("***tok: {}, top: {}, operators: {}, operands: {}".format(tok, top, self.operators, self.operands))
                    #if we haven't yet seen the other operand we need to hop out
                    if len(self.operands.elements) == 1:
                        break
                    
                    top = self.operators.pop()
                    if tok.get_precedence() > top.get_precedence():
                        break
                    
                    v2 = self.operands.pop()
                    v1 = self.operands.pop()
                    if top.att == '*':
                        self.operands.push(v1 * v2)
                    elif top.att == '/':
                        self.operands.push(v1 / v2)
                    elif top.att == '+':
                        self.operands.push(v1 + v2)
                    elif top.att == '-':
                        self.operands.push(v1 - v2)
                    else:
                        print("Error in shunting yard regular while loop")
                        sys.exit(0)

                    print("immediately after1: opoerators: {}, operands: {}".format(self.operators, self.operands))
                print("immediately after2: opoerators: {}, operands: {}".format(self.operators, self.operands))
                self.operators.push(tok)
            else:
                print("God only knows (in shunting yard)", tok)

            tok = sy_lexer.scan()
"""      
                    

"""
lexer = Lexer("test_source.txt")

while True:
    tok = lexer.scan()
    print(tok)
    if tok.token == "EOF":
        break
"""

parser = Parser("test_source.txt")
parser.parse()





